File: android/src/main/java/com/smileidentity/react/expo/SmileIDExpoModule.kt
package com.smileidentity.react.expo

import androidx.core.net.toUri
import com.smileidentity.SmileID
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import com.smileidentity.metadata.models.WrapperSdkName
import expo.modules.kotlin.exception.CodedException
import java.net.URL

class SmileIDExpoModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("SmileIDExpo")

        // functions
        AsyncFunction("initialize") Coroutine { useSandBox: Boolean, enableCrashReporting: Boolean, config: SmileConfig?, apiKey: String? ->
            val context = appContext.reactContext
                ?: throw IllegalStateException("Context is not available")

            withContext(Dispatchers.IO) {
                SmileID.setWrapperInfo(WrapperSdkName.ReactNativeExpo, "11.2.7")

                when {
                    // Case 1: Initialize with API key and config
                    apiKey != null && config != null -> {

                        val result = SmileID.initialize(
                            context = context,
                            config = config.toConfig(),
                            useSandbox = useSandBox,
                            enableCrashReporting = enableCrashReporting,
                            apiKey = apiKey
                        ).await()

                        result.getOrThrow()
                    }
                    // Case 2: Initialize with just config
                    config != null -> {
                        val result = SmileID.initialize(
                            context = context,
                            config = config.toConfig(),
                            useSandbox = useSandBox,
                            enableCrashReporting = enableCrashReporting,
                        ).await()

                        result.getOrThrow()
                    }
                    // Case 3: Basic initialization
                    else -> {
                        val result = SmileID.initialize(
                            context = context,
                            useSandbox = useSandBox,
                        ).await()
                        result.getOrThrow()
                    }
                }
            }
        }

        AsyncFunction("setCallbackUrl") { callbackUrl : String ->
            return@AsyncFunction SmileID.setCallbackUrl(URL(callbackUrl))

        }
        AsyncFunction("setAllowOfflineMode") { allowOfflineMode : Boolean ->
            return@AsyncFunction SmileID.setAllowOfflineMode(allowOfflineMode)
        }
        AsyncFunction("submitJob") { jobId : String ->
           try {
               val result = SmileID.submitJob(jobId)
               return@AsyncFunction result
           } catch (e: Exception) {
               throw CodedException(
                   "SubmitJobError",
                   "Failed to submit job: ${e.message}",
                   e
               )
           }
        }
        AsyncFunction("getSubmittedJobs") {
            try {
                val jobs = SmileID.getSubmittedJobs()
                return@AsyncFunction jobs
            } catch (e: Exception) {
                throw CodedException(
                    "GetSubmittedJobsError",
                    "Failed to get submitted jobs: ${e.message}",
                    e
                )
            }
        }

        AsyncFunction("getUnsubmittedJobs") {
            try {
                val jobs = SmileID.getUnsubmittedJobs()
                 return@AsyncFunction jobs
            } catch (e: Exception) {
                throw CodedException(
                    "GetUnsubmittedJobsError",
                    "Failed to get unsubmitted jobs: ${e.message}",
                    e
                )
            }
        }
        AsyncFunction("cleanup") { jobId : String ->
            try {
                val result  = SmileID.cleanup(jobId)
                return@AsyncFunction result
            } catch (e: Exception) {
                throw CodedException(
                    "CleanupError",
                    "Failed to cleanup job: ${e.message}",
                    e
                )
            }
        }

        // Views
        View(SmileIDDocumentVerificationView::class) {
            Events("onResult", "onError")
            Prop("params") { view: SmileIDDocumentVerificationView, config: DocumentVerificationParams ->
                view.updateConfig(config)
            }
        }

        View(SmileIDDocumentVerificationEnhancedView::class) {
            Events("onResult", "onError")
            Prop("params") { view: SmileIDDocumentVerificationEnhancedView, config: EnhancedDocumentVerificationParams ->
                view.updateConfig(config)
            }
        }

        View(SmileIDSmartSelfieEnrollmentView::class) {
            Events("onResult", "onError")
            Prop("params") { view: SmileIDSmartSelfieEnrollmentView, config: SmartSelfieParams ->
                view.updateConfig(config)
            }
        }

        View(SmileIDSmartSelfieEnrollmentEnhancedView::class) {
            Events("onResult", "onError")
            Prop("params") { view: SmileIDSmartSelfieEnrollmentEnhancedView, config: SmartSelfieParams ->
                view.updateConfig(config)
            }
        }

        View(SmileIDSmartSelfieAuthenticationView::class) {
            Events("onResult", "onError")
            Prop("params") { view: SmileIDSmartSelfieAuthenticationView, config: SmartSelfieParams ->
                view.updateConfig(config)
            }
        }

        View(SmileIDSmartSelfieAuthenticationEnhancedView::class) {
            Events("onResult", "onError")
            Prop("params") { view: SmileIDSmartSelfieAuthenticationEnhancedView, config: SmartSelfieParams ->
                view.updateConfig(config)
            }
        }

        View(SmileIDBiometricKYCView::class) {
            Events("onResult", "onError")
            Prop("params") { view: SmileIDBiometricKYCView, config: BiometricKYCParams ->
                view.updateConfig(config)
            }
        }
    }
}

File: ios/SmileIDExpoModule.swift
import ExpoModulesCore
import SmileID
import SwiftUI
import UIKit

public class SmileIDExpoModule: Module {

    public func definition() -> ModuleDefinition {

        Name("SmileIDExpo")

       AsyncFunction("initialize") {
           (
               useSandBox: Bool,
               enableCrashReporting: Bool,
               config: SmileConfig?,
               apiKey: String?
           ) async throws -> Void in
				 SmileID.setWrapperInfo(name: .reactNativeExpo, version: "11.2.7")

           if let apiKey = apiKey, let config = config {
               // Case 1: Initialize with API key and config
              SmileID.initialize(
                apiKey: apiKey,
                config: Config(
                    partnerId: config.partnerId,
                    authToken: config.authToken,
                    prodLambdaUrl: config.prodLambdaUrl,
                    testLambdaUrl: config.testLambdaUrl
                ),
                useSandbox: useSandBox
               )
           } else if let config = config {
               // Case 2: Initialize with just config
               SmileID.initialize(
                   config: Config(
                       partnerId: config.partnerId,
                       authToken: config.authToken,
                       prodLambdaUrl: config.prodLambdaUrl,
                       testLambdaUrl: config.testLambdaUrl
                   ),
                   useSandbox: useSandBox
               )
           } else {
               // Case 3: Basic initialization
               SmileID.initialize(useSandbox: useSandBox)
           }
       }

        // Set callback url
        AsyncFunction("setCallbackUrl") { (callbackUrl: String) async throws -> Void in
            guard let url = URL(string: callbackUrl) else {
                   throw Exception(name: "InvalidUrl", description: "Invalid callback URL: \(callbackUrl)")
                 }
                 SmileID.setCallbackUrl(url: url)
        }

        // Set offline mode
        AsyncFunction("setAllowOfflineMode") { (allowOfflineMode: Bool) async throws -> Void in
            SmileID.setAllowOfflineMode(allowOfflineMode: allowOfflineMode)
        }

        // Submit a job
        AsyncFunction("submitJob") { (jobId: String) async throws -> Void in
            do {
                try  SmileID.submitJob(jobId: jobId)
            } catch {
                throw error
            }
        }

        // Get submitted jobs
        AsyncFunction("getSubmittedJobs") { () async throws -> [String] in
            let submittedJobs: [String] =  SmileID.getSubmittedJobs()
            return submittedJobs
        }

        // Get unsubmitted jobs
        AsyncFunction("getUnsubmittedJobs") { () async throws -> [String] in
            let unsubmittedJobs: [String] = SmileID.getUnsubmittedJobs()
                return unsubmittedJobs
        }

        //  Cleanup job data
        AsyncFunction("cleanup") { (jobId: String) async throws -> Void in
            do {
                try  SmileID.cleanup(jobId: jobId)
            } catch {
                throw error
            }
        }

        // Document Verification View
        View(SmileIDDocumentVerificationView.self) {
            Events("onResult", "onError")

            Prop("params") { (
                view: SmileIDDocumentVerificationView,
                config: DocumentVerificationParams
            ) in
                view.updateConfig(config)
            }
        }

        // Enhanced Document Verification View
        View(SmileIDDocumentVerificationEnhancedView.self) {
            Events("onResult", "onError")

            Prop("params") { (
                view: SmileIDDocumentVerificationEnhancedView,
                config: EnhancedDocumentVerificationParams
            ) in
                view.updateConfig(config)
            }
        }

        // SmartSelfie Enrollment View
        View(SmileIDSmartSelfieEnrollmentView.self) {
            Events("onResult", "onError")

            Prop("params") { (
                view: SmileIDSmartSelfieEnrollmentView,
                config: SmartSelfieParams
            ) in
                view.updateConfig(config)
            }
        }

        // Enhanced SmartSelfie Enrollment View
        View(SmileIDSmartSelfieEnrollmentEnhancedView.self) {
            Events("onResult", "onError")

            Prop("params") { (
                view: SmileIDSmartSelfieEnrollmentEnhancedView,
                config: SmartSelfieParams
            ) in
                view.updateConfig(config)
            }
        }

        // SmartSelfie Authentication View
        View(SmileIDSmartSelfieAuthenticationView.self) {
            Events("onResult", "onError")

            Prop("params") { (
                view: SmileIDSmartSelfieAuthenticationView,
                config: SmartSelfieParams
            ) in
                view.updateConfig(config)
            }
        }

        // Enhanced SmartSelfie Authentication View
        View(SmileIDSmartSelfieAuthenticationEnhancedView.self) {
            Events("onResult", "onError")

            Prop("params") { (
                view: SmileIDSmartSelfieAuthenticationEnhancedView,
                config: SmartSelfieParams
            ) in
                view.updateConfig(config)
            }
        }

        // Biometric KYC View
        View(SmileIDBiometricKYCView.self) {
            Events("onResult", "onError")

            Prop("params") { (
                view: SmileIDBiometricKYCView,
                config: BiometricKYCParams
            ) in
                view.updateConfig(config)
            }
        }

    }
}
